### ISSUE TITLE
CSL?

### ISSUE THREAD

**Issue by bdarcus** [typst/hayagriva#32]
I'm admittedly biased, given I created it, but have you considered (also) supporting CSL citation styles and JSON input format?

I'm sure there are performance and other advantages to the rust-based styles, but there are thousands of CSL styles, as well as [citeproc-rs](https://github.com/zotero/citeproc-rs).

**Comment by hcsch** [comment]
Perhaps TOML (or similar more simple languages, like JSON, as mentioned above) would also be a nice alternative to YAML for this use case, since YAML is known to be rather unintuitive/counter-intuitive in too many cases (see https://noyaml.com/ for example).

**Comment by kmaasrud** [comment]
Using a well-defined format like CSL JSON, you'd easily be able to create a data structure that can be serialized/deserialized from any input format. This is possible already with the `citeproc-rs` crate. 

I agree with you @bdarcus, `hayagriva` really should leverage CSL, as the big contendors in bibliography management (notably Zotero) are fully in on it. 

@reknih, you should definitely hit up @cormacrelf and try to leverage his work wherever possible. The Rust ecosystem would definitely benefit from a proper citation processing library---I know that at least me and likely [the Tectonic guys](https://github.com/tectonic-typesetting) are interested!

**Comment by bdarcus** [comment]
I dunno; I think toml would be a poor format for this purpose, and YAML is actually pretty good. Among other things, you can validate it's file using JSON schemas.

**Comment by clbarnes** [comment]
citeproc-rs is possibly dead; readme calls it WIP, maintainer seems to have been active in the last couple of months but no commits to master since 2021.

**Comment by kmaasrud** [comment]
> citeproc-rs is possibly dead; readme calls it WIP, maintainer seems to have been active in the last couple of months but no commits to master since 2021. 

Not very active, no, but it's still maintained, as evidenced by the PR opened 2 months ago that fixes a bug introduced by Rust 1.67. Also, it being part of the Zotero org, one would think it has some official weight.

That being said, Hayagriva is way more elegant IMO, and it actually exists on crates.io. Adapting to support CSL JSON shouldn't be that difficult (I am considering opening a PR for it.) CSL styles is another beast, though, but one that this library should definitely aim for supporting!

**Comment by bdarcus** [comment]
[News](https://github.com/zotero/citeproc-rs/issues/61#issuecomment-1483791099) on `citeproc-rs`.

Basically they're stalled, with labor and technical hurdles, and need help to get the code in shape and released.

Another third-party developer is going to spend some time trying to figure if and how to do that.

On another though related note, I had an early morning, potentially crazy, idea of a simpler next-gen CSL that would use YAML and a generic template language, with fragments potentially something like.

```yaml
citation:
  mode: intext
  sort: false
  template: ({{ author.name | format-name-apa }}, {{ date | date-year }})
```

In liquid, the `|` delimits chained filters.

As I say, it might be a terrible idea, because there are a lot of details to consider, but I did post it on [the thread](https://discourse.citationstyles.org/t/upcoming-csl-meetup-context/1767/21) I linked to above to what people think.

**Comment by reknih** [comment]
Hi folks! I have already considered adding CSL, it's definitely on the roadmap!

It would be nice if I did not have to reimplement a Rust parser for CSL, is `citeproc-rs` up to the task?

**Comment by bdarcus** [comment]
> It would be nice if I did not have to reimplement a Rust parser for CSL, is `citeproc-rs` up to the task?

IDK; "csl" is one of only two crates he actually released.

Parsing is easy; it's just XML after all.

It's the processing that's difficult.

**Comment by bdarcus** [comment]
Dan posted [another](https://github.com/zotero/citeproc-rs/issues/61#issuecomment-1483994050) more detailed follow-up on the technical status.

**Comment by bdarcus** [comment]
FWIW, I've been playing with an experimental evolution of CSL in a [typescript model](https://github.com/bdarcus/csl-next.js); a [commented YAML file](https://github.com/bdarcus/csl-next.js/blob/main/examples/style.csl.yaml) of the current state.

As I say in the README, have no idea if this goes anywhere or not.

**Comment by bdarcus** [comment]
@reknih when you and/or your other developers have a bit free time, can you take a look at this?

https://github.com/bdarcus/csln

It's a reimplementation of the csl-next draft model in pure Rust, with very tight coupling  (thanks to serde) between the JSON schema input and internal model.

I'm pretty confident in that model, though it would need more review, testing, and iteration for me to be fully happy with it.

I'm much less confident in my programming skills, and the fact I'm a complete Rust newbie.

But I'm absolutely serious about building this out. I just need some help.

It should compile fine using the cargo, and I have it licensed under Mozilla 2.0, which I think should be compatible with your Apache option; probably not MIT. But my view on licenses is as a practical open source advocate. I choose the license simply because it's the same as citeproc-rs,

It's not quite pare with the typescript processor; here's an example of where I'm at:

```console
❯ target/debug/csln processor/examples/style.csl.yaml processor/examples/ex1.bib.yaml
```
Example result:

```rs
{
  "smith1": {
    "disamb-condition": false,
    "group-index": 1,
    "group-length": 1,
    "group-key": "Smith, Sam:2023-10"
  },
```

So the cored of the processor at this point is a sorted bibiliography vector, and this HashMap.

The next step is a function to iterate through the former and use the latter to generate the pre-rendered AST.

**Comment by reknih** [comment]
Hey @bdarcus, I recently started a CSL 1.0.2 XML parser and processor with typst/citationberg. Good to know that you are working on something for the next generation of CSL! What kind of help are you looking for?

**Comment by bdarcus** [comment]
@reknih 

> I recently started a CSL 1.0.2 XML parser and processor with typst/citationberg.

Oh, cool; didn't know!

How are you finding working with the XML?

> What kind of help are you looking for?

I hadn't gotten around yet (since this is newer) to sketching out milestones, but the ones for the typscript project more-or-less apply.

https://github.com/bdarcus/csl-next/milestones?direction=asc&sort=due_date

There's still a lot of work to do on the processor, for example, and we need to figure out a way to convert 1.0 styles, which may or may not be a big task.

It may be useful for you to review the model now, and think about whether there's promise in using that, and simply converting 1.0 styles to it, if we can do is fairly losslessly?

And perhaps there's a way to share code between them?

This is admittedly not fully-developed at this point, but I think I've thought-through enough details that it should work out as I intend.

On a more mundane level, since I'm a mediocre amateur programmer and rust newbie, reviews of existing code and PRs to improve would be welcome :-)

**Comment by fredguth** [comment]
Any news on that?  Still needs help?  What kind of processing is needed? What makes this task difficult? (Rush newbie)

**Comment by reknih** [comment]
Thank you for asking. The following tasks are still open:

- [ ] I am rewriting the frontend code for Hayagriva as well. Currently, I am moving the serialization and deserialization to `serde` which requires some manual trait impls. I aim to complete this today on a private branch and merge it with `new-frontend`.
- [ ] The `new-frontend` branch needs to be merged with `main` and `csl`
- [ ] On the `csl` branch, we need to translate CSL variables into Hayagriva field accesses. If need be, we need to introduce new fields.
- [ ] The `ChunkedStr` needs to be integrated with the case folding in the `lang` module.
- [ ] We need to fix and apply [_inheritable name options_](https://docs.citationstyles.org/en/stable/specification.html#inheritable-name-options) in both [citationberg](https://github.com/typst/citationberg) and Hayagriva, with a new context stack.
- [ ] We need to generate citations and bibliographies with multiple elements, sort and disambiguate them according to the CSL spec.
- [ ] We need to appropriately handle locales, including suppressing Title and Sentence case folding for non-English locales.
- [ ] [`page-range-format`](https://docs.citationstyles.org/en/stable/specification.html#global-options), [Citation-specific options](https://docs.citationstyles.org/en/stable/specification.html#toc-entry-50), and [Bibliography-specific options](https://docs.citationstyles.org/en/stable/specification.html#bibliography-specific-options) need to be implemented

**Comment by bdarcus** [comment]
What will be relationship between hayagriva and citationberg?

**Comment by reknih** [comment]
Citationberg parses CSL but makes no assumptions about how variables and data types are expressed within the consumer. Hayagriva will have a frontend to enter bibliographic information and be the CSL processor.

**Comment by reknih** [comment]
This has been shipped with 0.4.0
